File: 프로그래머스/lv2/81302. 거리두기 확인하기/거리두기 확인하기.cpp
#include <string>
#include <vector>
#include <queue>
using namespace std;

int dy[4] = {0,0,-1,1};
int dx[4] = {-1,1,0,0};

vector<int> solution(vector<vector<string>> places) {
    vector<int> answer;
    for (int i = 0; i < places.size(); i++)
    {
        bool flag = 0;
        for (int y = 0; y < 5; y++)
        {
            for (int x = 0; x < 5; x++)
            {
                if (places[i][y][x] == 'P')
                {
                    bool checked[5][5] = {0,};
                    checked[y][x] = 1;
                    queue <pair<pair<int,int>,int>> q;
                    q.push({{y,x},0});
                    while (!q.empty())
                    {
                        pair<pair<int,int>,int> temp = q.front();
                        q.pop();
                        if (temp.second == 2) continue;
                        for (int j = 0; j < 4; j++)
                        {
                            int ny = temp.first.first + dy[j];
                            int nx = temp.first.second + dx[j];
                            if (ny < 0 || nx < 0 || ny >= 5 || nx >= 5) continue;
                            if (checked[ny][nx] || places[i][ny][nx] == 'X') continue;
                            if (places[i][ny][nx] == 'P')
                            {
                                flag = 1;
                                break;
                            }
                            checked[ny][nx] = 1;
                            q.push({{ny,nx}, temp.second + 1});
                        }
                    }
                }
                if (flag) break;
            }
            if (flag) break;
        }
        if (flag) answer.push_back(0);
        else answer.push_back(1);
    }
    return answer;
}

File: 프로그래머스/2/81302. 거리두기 확인하기/거리두기 확인하기.java
import java.util.*;

class Solution {
    
    public int[] solution(String[][] places) {
        
        int [] answer = new int [places.length];
        int[][] dir = {
            {0, -1},
            {1, 0},
            {0, 1},
            {-1, 0}
        };
        
        for (int p = 0; p < places.length; p++) {
            String [] place = places[p];
            
            int flag = 1;
            for (int i = 0; i < 5; i++) {
                if (flag == 0) break;
                for (int j = 0; j < 5; j++) {
                    if (place[i].charAt(j) != 'P') continue;
                    if (flag == 0) break;
                    for (int k = 0; k < 4; k++) {
                        if (i + dir[k][0] < 0 || i + dir[k][0] >= 5 || j + dir[k][1] < 0 || j + dir[k][1] >= 5) {
                            continue;
                        }
                        if (place[i + dir[k][0]].charAt(j + dir[k][1]) == 'X') {
                            continue;
                        }
                        if (place[i + dir[k][0]].charAt(j + dir[k][1]) == 'P') {
                            flag = 0;
                            break;
                        }
                        else {
                            int x = i + dir[k][0];
                            int y = j + dir[k][1];
                            for (int t = 0; t < 4; t++) {
                                if (x + dir[t][0] < 0 || x + dir[t][0] >= 5 || y + dir[t][1] < 0 || y + dir[t][1] >= 5) {
                                    continue;
                                }
                                if (x + dir[t][0] == i && y + dir[t][1] == j)
                                {
                                    continue;
                                }
                                if (place[x + dir[t][0]].charAt(y + dir[t][1]) == 'P') {
                                    flag = 0;
                                    break;
                                }
                            }
                        }
                    }
                }
            }
            
            answer[p] = flag;
        }
        
        return answer;
    }
}